src/v/compression/lz4_decompression_buffers.cc

```diff
@@ -74,4 +75,134 @@ size_t lz4_decompression_buffers::min_alloc_threshold() const {
     return _min_alloc_threshold;
 }
 
+LZ4F_CustomMem lz4_decompression_buffers::custom_mem_alloc() {
+    // If custom allocation is disabled, setting all alloc functions to null
+    // makes lz4 fall back to malloc, calloc and free.
+    if (_disabled) {
+        return {
+          .customAlloc = nullptr,
+          .customCalloc = nullptr,
+          .customFree = nullptr,
+          .opaqueState = nullptr};
+    }
+
+    return {
+      .customAlloc = alloc_lz4_obj,
+      .customCalloc = nullptr,
+      .customFree = free_lz4_obj,
+      .opaqueState = this};
+}
+
 } // namespace compression
+
+// During a typical lz4 decompression operation the following LZ4F_malloc calls
+// will be processed via this alloc function:
+// 1. Allocation for the tmp input buffer: this can be a maximum of 4MiB + 4
+// bytes
+// 2. Allocation for the tmp output buffer: this can be a maximum of 4MiB +
+// 128KiB
+// These two calls will typically happen once per decompression context, and are
+// preceded by calls to LZ4F_free to first free up the two buffers.
+void* alloc_lz4_obj(void* state, size_t size) {
+    auto* st = static_cast<compression::lz4_decompression_buffers*>(state);
+    vassert(
+      size <= st->buffer_size(),
+      "Request to allocate {} bytes which is more than max buffer size "
+      "available: {} bytes",
+      size,
+      st->buffer_size());
+
+    if (size < st->min_alloc_threshold()) {
+        st->pass_through_allocated();
+        return malloc(size);
+    }
+
+    vassert(
+      st->has_reserved_buffers(),
+      "must reserve space with lz4_decompression_buffers::reserve_buffers() "
+      "before allocation");
+
+    auto& bufs = st->buffers();
```

Comment:
if we already have a chunk of 4mb memory, wouldn't using `malloc` be strictly more work than using the arena? what's the advantage to allocating more memory when its small?

Reply:
When we opt in to use this custom memory manager with lz4 it will route all calls to malloc and free via our provided functions. So if there are any calls to malloc made by lz4 during decompression other than the ones for the two buffers we care about, those should still be serviceable. The idea behind falling back to malloc for small buffers is to service those calls. We don't want to use the preallocated buffers for those calls.

On the other hand I looked through the usages for `LZ4F_malloc` and did not find any calls to it other than those two buffers in the decompression code path. So it might work but I'm not sure if it is guaranteed that calls with remain the same within lz4.

Reply:
>When we opt in to use this custom memory manager with lz4 it will route all calls to malloc and free via our provided functions. So if there are any calls to malloc made by lz4 during decompression other than the ones for the two buffers we care about, those should still be serviceable. The idea behind falling back to malloc for small buffers is to service those calls. We don't want to use the preallocated buffers for those calls.

I see. This seems very dangerous, right?

> We don't want to use the preallocated buffers for those calls.

Specifically, you can't use preallocated buffers for those calls unless there is exactly one such call, right? IIUC you are depending on lz4 making arbitrary malloc/free calls, where _exactly one of those malloc/free calls_ is a call that have exclusive access to the pre-allocated region?

Reply:
It should be safe. Right now we can describe the calls to `malloc` during decompression as below:

* all calls to `malloc` <= `128_KiB` go to the system `malloc` implementation
* all calls to `malloc` > `128_KiB` go to our two buffers, which are allocated in sequence. 

We know from the source code of LZ4 that during decompression it makes exactly two large sized `malloc` calls, one for input buffer and one for output buffer. 

Both are relatively similar in size, and depend on block size. So they can be 64 KiB, 256 KiB, 1 MiB or 4 MiB each. In all but the first case our buffers will be allocated out. For all other smaller sized `malloc` calls during decompression system `malloc` will be used.

Note that the `malloc` calls are made when the decompression context is in the init stage, which only happens once per a review of the decompression code path.

If we look at the free calls they are a lot simpler:

* For addresses belonging to our two preallocated buffers, we just update the state in our free implementation.
* For all other addresses we fall back to the system `free` impl.

Also if the lz4 library changes in future and more `malloc`s are added, our unit test will start failing because we have exact call counts in the decompression code path test.

Reply:
>  IIUC you are depending on lz4 making arbitrary malloc/free calls, where exactly one of those malloc/free calls is a call that have exclusive access to the pre-allocated region?

Yes, there are two calls to malloc expected which will use our buffers. All others should go to system malloc. In the current lz4 implementation, during decompression, there are only two malloc calls. These are expected to result in allocation of our two buffers.